File: Python/GLA.py
#Author Name: Laxmi Chari
#Roll No: 22
#Title: Program to implement OO Concepts using Python
#Start Date: 22/07/2024,
#Modified Date: 25/07/2024
#Description: The program helps in understanding the structure of the Goa Legislative Assembly using Object Oriented Concepts 

# Abstract base class representing a generic entity in the legislative assembly
class Entity:
    def __str__(self):
        pass  # Abstract method to be implemented by subclasses for string representation


# Class representing a legislative bill
class Bill(Entity):
    def __init__(self, title, description):
        self.title = title  # Title of the bill
        self.description = description  # Description of the bill
        self.is_passed = False  # Indicates whether the bill has been passed

    # Method to mark the bill as passed
    def pass_bill(self):
        self.is_passed = True  # Set the is_passed attribute to True

    # Override the __str__ method to provide a string representation of the bill
    def __str__(self):
        return f"Title: {self.title}\nDescription: {self.description}\nStatus: {'Passed' if self.is_passed else 'Not Passed'}"


# Class representing a Member of Legislative Assembly (MLA)
class MLA(Entity):
    def __init__(self, name, constituency, party):
        self.name = name  # Name of the MLA
        self.constituency = constituency  # Constituency of the MLA
        self.party = party  # Political party of the MLA

    # Override the __str__ method to provide a string representation of the MLA
    def __str__(self):
        return f"Name: {self.name}\nConstituency: {self.constituency}\nPolitical party: {self.party}"


# Class representing an official in the assembly (e.g., Speaker, Deputy Speaker, Governor)
class Official(Entity):
    def __init__(self, position, name):
        self.position = position  # Position of the official (e.g., Speaker)
        self.name = name  # Name of the official

    # Override the __str__ method to provide a string representation of the official
    def __str__(self):
        return f"Position: {self.position}\nName: {self.name}"


# Class representing a session of the legislative assembly
class Session(Entity):
    def __init__(self, date, agenda):
        self.date = date  # Date of the session
        self.agenda = agenda  # Agenda of the session

    # Override the __str__ method to provide a string representation of the session
    def __str__(self):
        return f"Date of the session: {self.date}\nAgenda: {self.agenda}"


# Class to manage the Goa Legislative Assembly
class GoaLegislativeAssembly:
    def __init__(self):
        # Lists to store sessions, bills, officials, and MLAs
        self.sessions = []
        self.bills = []
        self.officials = []
        self.mlas = []

    # Method to display the menu and handle user input
    def menu(self):
        while True:
            # Display menu options
            print("\nGoa Legislative Assembly Menu")
            print("1. Enter the Session Details")
            print("2. Display Session Details")
            print("3. Introduce a Bill")
            print("4. Pass a Bill")
            print("5. Display Bills")
            print("6. Add an Official")
            print("7. View Officials")
            print("8. Add an MLA")
            print("9. View MLAs")
            print("10. Exit")
            choice = int(input("Enter your choice: "))  # Read user choice

            # Handle user choice using if-elif statements
            if choice == 1:
                self.session_details()  # Call method to enter session details
            elif choice == 2:
                self.display_session()  # Call method to display session details
            elif choice == 3:
                self.introduce_bill()  # Call method to introduce a new bill
            elif choice == 4:
                self.pass_bill()  # Call method to pass a bill
            elif choice == 5:
                self.display_bills()  # Call method to display all bills
            elif choice == 6:
                self.add_official()  # Call method to add an official
            elif choice == 7:
                self.view_officials()  # Call method to view all officials
            elif choice == 8:
                self.add_mla()  # Call method to add a new MLA
            elif choice == 9:
                self.view_mlas()  # Call method to view all MLAs
            elif choice == 10:
                print("Exiting...")  # Exit message
                break  # Exit the loop
            else:
                print("Invalid choice. Please try again.")  # Handle invalid choices

    # Method to enter session details
    def session_details(self):
        date = input("Enter the Date of the session: ")  # Read session date
        agenda = input("Enter the Agenda of the session: ")  # Read session agenda
        self.sessions.append(Session(date, agenda))  # Add the new session to the list
        print("Session may proceed.")  # Confirmation message

    # Method to display details of all sessions
    def display_session(self):
        if not self.sessions:  # Check if there are no sessions
            print("No session conducted yet.")  # Handle case where no sessions exist
            return

        # Loop through and display all sessions
        for session in self.sessions:
            print(session)  # Print session details
            print()
        
        if not self.bills:  # Check if there are no bills
            print("No bills introduced in the session.")  # Handle case where no bills exist
        else:
            print("Bills introduced in the session:")
            for bill in self.bills:  # Loop through bills to display them
                print(bill)  # Print bill details
                print()
        
        if not self.officials:  # Check if there are no officials
            print("No officials to display.")  # Handle case where no officials exist
        else:
            print("Officials present in today's Session:")
            for official in self.officials:  # Loop through officials to display them
                print(official)  # Print official details
                print()
        
        if not self.mlas:  # Check if there are no MLAs
            print("No MLAs to display.")  # Handle case where no MLAs exist
        else:
            print("List of MLAs present for today's session:")
            for mla in self.mlas:  # Loop through MLAs to display them
                print(mla)  # Print MLA details
                print()

    # Method to introduce a new bill
    def introduce_bill(self):
        title = input("Enter the title of the bill: ")  # Read bill title
        description = input("Enter the description of the bill: ")  # Read bill description
        self.bills.append(Bill(title, description))  # Add the new bill to the list
        print("Bill introduced successfully.")  # Confirmation message

    # Method to pass a bill
    def pass_bill(self):
        if not self.bills:  # Check if there are no bills
            print("No bills available to pass.")  # Handle case where no bills exist
            return
        
        title = input("Enter the title of the bill to pass: ")  # Read the title of the bill to pass
        # Loop through bills to find the one matching the title
        for bill in self.bills:
            if bill.title == title:  # Check if bill title matches
                bill.pass_bill()  # Mark the bill as passed
                print("Bill passed successfully.")  # Confirmation message
                return
        print("Bill not found.")  # Handle case where the bill is not found

    # Method to display all bills
    def display_bills(self):
        if not self.bills:  # Check if there are no bills
            print("No bills to display.")  # Handle case where no bills exist
            return
        
        # Loop through and display all bills
        for bill in self.bills:
            print(bill)  # Print bill details
            print()

    # Method to add a new official
    def add_official(self):
        position = input("Enter the position of the official: ")  # Read official's position
        name = input("Enter the name of the official: ")  # Read official's name
        self.officials.append(Official(position, name))  # Add the new official to the list
        print("Official added successfully.")  # Confirmation message

    # Method to view all officials
    def view_officials(self):
        if not self.officials:  # Check if there are no officials
            print("No officials to display.")  # Handle case where no officials exist
            return
        
        print("List of Officials:")  # Display header
        # Loop through and display all officials
        for official in self.officials:
            print(official)  # Print official details
            print()

    # Method to add a new MLA
    def add_mla(self):
        name = input("Enter the name of the MLA: ")  # Read MLA name
        constituency = input("Enter the constituency of the MLA: ")  # Read MLA constituency
        party = input("Enter the Political party of the MLA: ")  # Read MLA party
        self.mlas.append(MLA(name, constituency, party))  # Add the new MLA to the list
        print("MLA added successfully.")  # Confirmation message

    # Method to view all MLAs
    def view_mlas(self):
        if not self.mlas:  # Check if there are no MLAs
            print("No MLAs to display.")  # Handle case where no MLAs exist
            return
        
        print("List of MLAs:")  # Display header
        # Loop through and display all MLAs
        for mla in self.mlas:
            print(mla)  # Print MLA details
            print()


# Main block to run the application
if __name__ == "__main__":
    assembly = GoaLegislativeAssembly()  # Create an instance of the GoaLegislativeAssembly class
    assembly.menu()  # Start the menu

File: Java/GLA.java
/* Author Name: Laxmi Chari
Roll No: 22
Title: Program to implement OO Concepts using JAVA
Start Date: 21/07/2024
Modified Date: 25/07/2024
Description: The program helps in understanding the structure of the Goa Legislative assembly using Object Oriented Concepts 
 */
// Abstract base class representing a generic entity in the legislative assembly
public abstract class Entity {
    // Abstract method that must be implemented by subclasses for string representation
    @Override
    public abstract String toString();
}

// Class representing a legislative bill
public class Bill extends Entity {
    private String title;        // Title of the bill
    private String description;  // Description of the bill
    private boolean isPassed;    // Indicates whether the bill has been passed

    // Constructor to initialize a new bill with a title and description
    public Bill(String title, String description) {
        this.title = title;
        this.description = description;
        this.isPassed = false;  // Bill is not passed initially
    }

    // Method to mark the bill as passed
    public void pass() {
        this.isPassed = true;
    }

    // Override the toString method to provide a string representation of the bill
    @Override
    public String toString() {
        return "Title: " + title + "\nDescription: " + description + "\nStatus: " + (isPassed ? "Passed" : "Not Passed");
    }
}

// Class representing a Member of Legislative Assembly (MLA)
public class MLA extends Entity {
    private String name;          // Name of the MLA
    private String constituency;  // Constituency of the MLA
    private String party;         // Political party of the MLA

    // Constructor to initialize an MLA with name, constituency, and party
    public MLA(String name, String constituency, String party) {
        this.name = name;
        this.constituency = constituency;
        this.party = party;
    }

    // Override the toString method to provide a string representation of the MLA
    @Override
    public String toString() {
        return "Name: " + name + "\nConstituency: " + constituency + "\nPolitical party: " + party;
    }
}

// Class representing an official in the assembly (e.g., Speaker, Deputy Speaker, Governor)
public class Official extends Entity {
    private String position;  // Position of the official (e.g., Speaker)
    private String name;      // Name of the official

    // Constructor to initialize the official's position and name
    public Official(String position, String name) {
        this.position = position;
        this.name = name;
    }

    // Override the toString method to provide a string representation of the official
    @Override
    public String toString() {
        return "Position: " + position + "\nName: " + name;
    }
}

// Class representing a session of the legislative assembly
public class Session extends Entity {
    private String date;    // Date of the session
    private String agenda;  // Agenda of the session

    // Constructor to initialize a session with date and agenda
    public Session(String date, String agenda) {
        this.date = date;
        this.agenda = agenda;
    }

    // Override the toString method to provide a string representation of the session
    @Override
    public String toString() {
        return "Date of the session: " + date + "\nAgenda: " + agenda;
    }
}

// Main class to manage the Goa Legislative Assembly
import java.util.ArrayList; // Importing the ArrayList class to manage collections
import java.util.Scanner;   // Importing the Scanner class for user input

public class GoaLegislativeAssembly {
    // Lists to store sessions, bills, officials, and MLAs
    private static ArrayList<Session> sessions = new ArrayList<>();
    private static ArrayList<Bill> bills = new ArrayList<>();
    private static ArrayList<Official> officials = new ArrayList<>();
    private static ArrayList<MLA> mlas = new ArrayList<>();
    private static Scanner scanner = new Scanner(System.in); // Scanner for reading user input

    public static void main(String[] args) {
        int choice; // Variable to store user menu choice
        do {
            // Display the menu options
            System.out.println("\nGoa Legislative Assembly Menu");
            System.out.println("1. Enter the Session Details");
            System.out.println("2. Display Session Details");
            System.out.println("3. Introduce a Bill");
            System.out.println("4. Pass a Bill");
            System.out.println("5. Display Bills");
            System.out.println("6. Add an Official");
            System.out.println("7. View Officials");
            System.out.println("8. Add an MLA");
            System.out.println("9. View MLAs");
            System.out.println("10. Exit");
            System.out.print("Enter your choice: ");
            choice = scanner.nextInt(); // Read user choice
            scanner.nextLine(); // Clear the buffer

            // Handle user choice using a switch statement
            switch (choice) {
                case 1:
                    sessionDetails();  // Call method to enter session details
                    break;
                case 2:
                    displaySession();  // Call method to display session details
                    break;
                case 3:
                    introduceBill();   // Call method to introduce a new bill
                    break;
                case 4:
                    passBill();       // Call method to pass a bill
                    break;
                case 5:
                    displayBills();    // Call method to display all bills
                    break;
                case 6:
                    addOfficial();     // Call method to add an official
                    break;
                case 7:
                    viewOfficials();    // Call method to view all officials
                    break;
                case 8:
                    addMLA();         // Call method to add a new MLA
                    break;
                case 9:
                    viewMLAs();       // Call method to view all MLAs
                    break;
                case 10:
                    System.out.println("Exiting..."); // Exit message
                    break;
                default:
                    System.out.println("Invalid choice. Please try again."); // Handle invalid choices
            }
        } while (choice != 10); // Continue until the user chooses to exit
    }

    // Method to enter session details
    private static void sessionDetails() {
        System.out.print("Enter the Date of the session: ");
        String date = scanner.nextLine(); // Read session date
        System.out.print("Enter the Agenda of the session: ");
        String agenda = scanner.nextLine(); // Read session agenda
        sessions.add(new Session(date, agenda)); // Add the new session to the list
        System.out.println("Session may proceed."); // Confirmation message
    }

    // Method to display details of all sessions
    private static void displaySession() {
        if (sessions.isEmpty()) {
            System.out.println("No session conducted yet."); // Handle case where no sessions exist
            return;
        }
        // Loop through and display all sessions
        for (Session session : sessions) {
            System.out.println(session); // Print session details
            System.out.println();
        }

        // Display bills introduced in the session
        System.out.println("Bills introduced in the session.");
        for (Bill bill : bills) {
            System.out.println(bill); // Print bill details
            System.out.println();
        }

        // Check if there are any officials to display
        if (officials.isEmpty()) {
            System.out.println("No officials to display."); // No officials found
            return;
        }
        // Display officials present in today's session
        System.out.println("Officials present in today's Session:");
        for (Official official : officials) {
            System.out.println(official); // Print official details
            System.out.println();
        }

        // Check if there are any MLAs to display
        if (mlas.isEmpty()) {
            System.out.println("No MLAs to display."); // No MLAs found
            return;
        }
        // Display MLAs present for today's session
        System.out.println("List of MLAs present for today's session:");
        for (MLA mla : mlas) {
            System.out.println(mla); // Print MLA details
            System.out.println();
        }
    }

    // Method to introduce a new bill
    private static void introduceBill() {
        System.out.print("Enter the title of the bill: ");
        String title = scanner.nextLine(); // Read bill title
        System.out.print("Enter the description of the bill: ");
        String description = scanner.nextLine(); // Read bill description
        bills.add(new Bill(title, description)); // Add the new bill to the list
        System.out.println("Bill introduced successfully."); // Confirmation message
    }

    // Method to pass a bill
    private static void passBill() {
        if (bills.isEmpty()) {
            System.out.println("No bills available to pass."); // Handle case where no bills exist
            return;
        }
        System.out.print("Enter the title of the bill to pass: ");
        String title = scanner.nextLine(); // Read the title of the bill to pass
        // Loop through bills to find the one matching the title
        for (Bill bill : bills) {
            if (bill.toString().contains(title)) { // Check if bill title matches
                bill.pass(); // Mark the bill as passed
                System.out.println("Bill passed successfully."); // Confirmation message
                return;
            }
        }
        System.out.println("Bill not found."); // Handle case where the bill is not found
    }

    // Method to display all bills
    private static void displayBills() {
        if (bills.isEmpty()) {
            System.out.println("No bills to display."); // Handle case where no bills exist
            return;
        }
        // Loop through and display all bills
        for (Bill bill : bills) {
            System.out.println(bill); // Print bill details
            System.out.println();
        }
    }

    // Method to add a new official
    private static void addOfficial() { // speaker or deputy speaker or governor
        System.out.print("Enter the position of the official: ");
        String position = scanner.nextLine(); // Read official's position
        System.out.print("Enter the name of the official: ");
        String name = scanner.nextLine(); // Read official's name
        officials.add(new Official(position, name)); // Add the new official to the list
        System.out.println("Official added successfully."); // Confirmation message
    }

    // Method to view all officials
    private static void viewOfficials() {
        if (officials.isEmpty()) {
            System.out.println("No officials to display."); // Handle case where no officials exist
            return;
        }
        System.out.println("List of Officials:"); // Display header
        // Loop through and display all officials
        for (Official official : officials) {
            System.out.println(official); // Print official details
            System.out.println();
        }
    }

    // Method to add a new MLA
    private static void addMLA() {
        System.out.print("Enter the name of the MLA: ");
        String name = scanner.nextLine(); // Read MLA name
        System.out.print("Enter the constituency of the MLA: ");
        String constituency = scanner.nextLine(); // Read MLA constituency
        System.out.print("Enter the Political party of the MLA: ");
        String party = scanner.nextLine(); // Read MLA party
        mlas.add(new MLA(name, constituency, party)); // Add the new MLA to the list
        System.out.println("MLA added successfully."); // Confirmation message
    }

    // Method to view all MLAs
    private static void viewMLAs() {
        if (mlas.isEmpty()) {
            System.out.println("No MLAs to display."); // Handle case where no MLAs exist
            return;
        }
        System.out.println("List of MLAs:"); // Display header
        // Loop through and display all MLAs
        for (MLA mla : mlas) {
            System.out.println(mla); // Print MLA details
            System.out.println();
        }
    }
}






